### PR TITLE
Feat: Issue#35 view user additional information

### LINKED ISSUE #35: View User Profile - Additional Information
## Description
As a registered user,
I want to see my profile - additional information,
so that I can update it in the future when necessary.

## Mocks
#5 

## Acceptance Criteria
- [ ] Additional profile information details - phone, mobile, number, personal website, etc. should be displayed
- [ ] Tests for display of each of these fields must be present

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentor.

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Allow user to view their additional infomation page

Fixes #35 

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
By login with as a user and geet additional information page by following the path:
 `My Space` -> `Profile` -> on `Personal Details` page hit `Continue` button and see below screenshots:
* if user is a representative of an organization:
Screen Shot 2020-07-17 at 3 48 04 pm

* if user is not a repreesentative of an organization:
Screen Shot 2020-07-17 at 3 48 48 pm

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] New and existing unit tests pass locally with my changes

**Additional Note**
This is a WIP. Still need to add tests and fix bugs

### CODE REVIEW

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . Here's the PR for View User's Additional Information page. This is still WIP as it doesn't have tests yet and still have minor bugs need to be fixed (to do with routing/navigating to previous page)

File: src/Enum.js
Comment (meenakshi-dhanani) [inline src/Enum.js]: there is no enum in javascript. There are objects that have key value pairs. You can simply call it timezones

Comment (mtreacy002) [reply src/Enum.js]: @meenakshi-dhanani , the reason I did this is because the backend respond to this request with the timezone enum `name` 
Screen Shot 2020-07-19 at 10 30 46 pm

which is what's being recorded in postgresql.
Screen Shot 2020-07-19 at 10 27 25 pm

Screen Shot 2020-07-19 at 10 32 15 pm
Screen Shot 2020-07-19 at 10 59 28 pm


However, what I want the user to see is the timezone `value` which according to [this stackoverflow discussion](https://stackoverflow.com/questions/18111657/how-to-get-names-of-enum-entries) could be done in in this way (what I've done in the code).

Screen Shot 2020-07-19 at 10 28 15 pm

The difference of the code are as per screenshoot below (the one commented out is returning timezone as its name, not value)
Screen Shot 2020-07-19 at 10 55 16 pm


There is another alternative, which is to change the backend response and request model to accept/return human-readable format (aka enum value) from the backend side and do the logic to convert it to enum name before saving it to the database. 
But again, seeing how we keep machine format (unix datetime) for datetime on our backend and only change it to human-readable on frontend, I thought there's a convention somewhere (I'm not sure if there is, but @ramitsawhney27 maybe can clarify this for us) to keep backend as close as machine language

Screen Shot 2020-07-19 at 10 31 12 pm

Comment (meenakshi-dhanani) [reply src/Enum.js]: Choose to keep the mapping logic in the frontend/backend. And follow it consistently. Based on our discussion, keeping it in the backend right now is less effort. So you can continue with that. So frontend will not keep any mappings, as per my understanding

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: Why the line?

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: can you make this data.message?

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx:123]: are we using snake case or camel case? 

additionalInfo = camel case
is_organization_rep = snake case

File: src/Enum.js
Comment (mtreacy002) [reply src/Enum.js]: Ok. will change it now.

File: src/myspace/AdditionalInfo.jsx
Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: I'll try

File: src/myspace/AdditionalInfo.jsx
Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: got it. now removed.

File: src/myspace/AdditionalInfo.jsx
Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx:123]: snake case on the response body of backend as per python convention. do you want the frontend to be changed as well (like in useState we use camel case)

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx:123]: @meenakshi-dhanani . I've replaced the snake case in any other place except setting value from dict additionalInfo as this is as per what being sent by backend (as python format)

**Comment by mtreacy002** [comment]
update @meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've pushed the latest chage. Please re-review.

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx:10]: We might want to persist the access_token as accessToken on our side. Because the backend might me using snake case, but on the frontend it's camel case everywhere. That would be a change in the context, I don't mind if you want to do that change separately. So I won't make that a blocker for this PR.
Small things, but consistent casing for the code hygiene is important

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: error message is already getting set in catch right? would there be a case that on error the control would not go to a catch? Just wondering?

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: @meenakshi-dhanani . For some reason, it seems like only the uncustomised errors get inside catch. where as the customised errors did not get inside catch. you can test this by running the app, login, but skip going to personal details page and go straight to additional info. with error also set inside get, you'll see the error message. but if you comment it out and just relying from the one inside catch, you'll not see the error message. I don't know why this happen, but maybe just the way `fetch` is, or perhaps something in backend need to be change. Any suggestions? cc @ramitsawhney27

File: src/myspace/AdditionalInfo.jsx
Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx:10]: Ok, can some else pick it up (like first-timer-only)? or do you want me to work on it?

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . I've added backend Heroku pr app server url to config. I may do the same on the next PR as well (not promising) so you can test frontend with the matching PR Heroku backend server instead of running locally

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [reply src/myspace/AdditionalInfo.jsx]: @mtreacy002 I checked fetch it works that way. Doesn't reject a 400. I'm looking at why and figuring another way. I'll post links on this soon.

Comment (meenakshi-dhanani) [reply src/myspace/AdditionalInfo.jsx]: I checked the documentation - https://javascript.info/fetch. Fetch doesn't reject http status errors. Usually how we handle it is since this same logic will be repeated for every call, we have a wrapper over api calls. You can push this and I can open another PR for that. However even in this, what you'd want to do it, check if the !response.ok and reject the promise with the data. Let catch then be the only place where you setErrorMessage. Once you reject anything, it will go to the catch

Comment (meenakshi-dhanani) [reply src/myspace/AdditionalInfo.jsx]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject - Inside the new error you can give the data from response.json

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: got it. made the change. will push now

**Comment by mtreacy002** [comment]
Update @meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've made the requested changes. Can you please re-review? Ta

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [reply src/myspace/AdditionalInfo.jsx:10]: We can create a separate one and label as first timer I think :)

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: this is unused

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: this const status is also not being used. You can remove this

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: oops.. forgot to remove it. Ok. done now. 😉

File: src/myspace/AdditionalInfo.jsx
Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx]: forgot this one too. Now it''s removed

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx:29]: if you throw an error, and send the data.message, in your catch, you will get access to that data.message. So you should be able to say setErrorMessage(error) 

while throwing simply throw new Error(data)

once you do that in catch you can say .catch(error) {set.... error.message}

Please note: whatever you send inside new Error will come to catch

Have you tried this out locally, when you run the app? You might want to check with a 500 scenario, and a 400 scenario both

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx:29]: Yep @meenakshi-dhanani . I tried it (with the existing code). The error is shown when user didn't retrieved their details before viewing additional info page.

ezgif com-video-to-gif (3)

Note that there's a backend bug in this application (already open [issue#94](https://github.com/anitab-org/bridge-in-tech-backend/issues/94) to deal with it) Logout on frontend currently only removes the cookies on frontend to allow toggling navbar with isAuth status. However, no logout being implemented on backend, thus logging out from frontend does not remove cookies of user_id in backend. At this stage, to remove this cookies on backend, the bit backend seerver needs to be terminated completely and ree-run again for frontend to login. 

See the behaviour below. User was logged-in, then logged-out, then login again. Expected behaviour is for user to be warned that they haven't retrieved their personal details. But instead, additional info is shown as if they had done so (retrieved their details)

ezgif com-video-to-gif (4)

Comment (mtreacy002) [reply src/myspace/AdditionalInfo.jsx:29]: Update @meenakshi-dhanani . I've just pushed the modified throw as per your suggestion. It still works the same as gif above

File: src/myspace/AdditionalInfo.jsx
Comment (meenakshi-dhanani) [inline src/myspace/AdditionalInfo.jsx]: we can look at this today in our call - https://www.geeksforgeeks.org/reject-vs-throw-promises-in-javascript/

**Review by meenakshi-dhanani — APPROVED** [review]

**Comment by meenakshi-dhanani** [comment]
@ramitsawhney27 @foongminwong can you approve if it looks fine? I reviewed it with Maya

**Review by foongminwong — APPROVED** [review]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses were tested as below:
    
    _PR 40_  
    _Screenshot/gif/url_:  
       
      test-pr-40

    
    _Expected Result_:  As a user, I should be able to view user additional information. If user additional information is not present, it should notify the user to click o the `Personal Details` page first.
    _Actual Result_: Same as expected.

3. OS Version: Windows 10

**Pull request merged**
